### PR TITLE
Added support PQC in the browser with the target wasm32-wasi now supported

### PR DESCRIPTION
**Pull request by john-sharratt**
As a part of a project I am working on there is a heavy use of quantum resistant crypto.

Now that I am porting the library (https://github.com/john-sharratt/ate) over to WebAssembly so that it can run in a web browser I need the same crypto functions to work in wasm32 under wasi however there are no known libraries in rust that support such routines other than pqcrypto - hence I started porting it.

Given that pqcrypto does not yet work I forked it to add support for compiling to wasm32-wasi (https://github.com/john-sharratt/pqcrypto-wasi) - most of it compiles except for the random generator which is runtime specific thus I have added a function within WASI.

In order to test simply run:
cargo build --no-default-features --target wasm32-wasi

Note: While many files are changed in this commit this is due to the nature of the code generation script - in reality the only files I changed where...

pqcrypto-internals/build.rs
pqcrypto-internals/include/*.h
pqcrypto-internals/cfiles/*.c
pqcrypto-templates/schema/build.rs.j2

The changes that were made was to support an external sys-root which must be supplied in the parent folder outside of the scope of this build.

Due to the changes you made with breaking the hard dependency on STD there are now no more changes required elsewhere upstream.

### CODE REVIEW

**Comment by thomwiggers** [comment]
I'm sorry, I had some commits that I'd used to prepare a release but forgot to push as I got called out to a meeting. Can you rebase?

I will try to make some time for testing after that.

**Comment by john-sharratt** [comment]
There's a problem on your master branch though...

./generate-implementations.py 
Traceback (most recent call last):
  File "./generate-implementations.py", line 180, in <module>
    generate_scheme(name, 'kem', properties)
  File "./generate-implementations.py", line 95, in generate_scheme
    metadatas[scheme['name']] = read_scheme_metadata(type, scheme['name'])
  File "./generate-implementations.py", line 22, in read_scheme_metadata
    with open(metadata_path) as f:
FileNotFoundError: [Errno 2] No such file or directory: 'pqclean/crypto_kem/ntrulpr953/META.yml'

**Comment by john-sharratt** [comment]
I'll do some testing in the mean-time

**Comment by john-sharratt** [comment]
It might actually be my fault - I'll work on it some more

**Comment by john-sharratt** [comment]
Ok - all fixed up now.

**Comment by john-sharratt** [comment]
Found a linking error on when of my test runs, this last commit should fix it.
Will continue my testing

**Comment by john-sharratt** [comment]
Screenshot from 2021-09-30 23-04-12

**Comment by john-sharratt** [comment]
confirmed that it compiles and works in browsers - awesome

**Comment by john-sharratt** [comment]
@thomwiggers I published the pqcrypto crates with a postfix "-wasi" and commit to delete them again - my project can't wait any more but whatever you want me to change in the code to get this merge finished... just let me know.

Willing to put the time in to help.

Thanks for the help so far

**Comment by thomwiggers** [comment]
This should also really be added to CI somehow.

**Comment by john-sharratt** [comment]
> This should also really be added to CI somehow.

I can write some bash scripts if you want me to show you how to test it?
The tool chain would have to be downloaded though so I'll split it into a script to install the tool chain and a script to run the tests

You want that in the merge request or just attached to this comment thread?

**Comment by john-sharratt** [comment]
I looked into setting up CI/CD for this but it was too much work in the end to make a special case for this project.
It will of course be picked up in my CI/CD for the ATE project but thats obviously after you publish.

If you want any help in the form of tips and questions/answers let me know.

**Comment by john-sharratt** [comment]
@thomwiggers - what do you want to do with this PR as I made the changes but you modified the build.rs file and I can't keep updating this pull request.

In the meantime I've got this all working by republishing the libraries but its not good to fork a WASI version from the mainline so its in the communities interests to sort this.

Cheers

**Comment by thomwiggers** [comment]
I'm afraid I just don't know enough about WASI's restrictions to comfortably merge this right now, without CI. It doesn't have to be complicated; just running the build and tests is probably enough.

**Comment by john-sharratt** [comment]
@thomwiggers  No worries - let me see what I can do

**Comment by john-sharratt** [comment]
There we go - all merge conflicts resolved and added some CI/CD that will download the SDK and build the crypto libraries against the WASI build target. This will find 99% of the errors as there really isn't that much else that can go wrong. Executing cargo wasi test will be too hard to get working as its not running the tests in rust.

See here:
https://github.com/john-sharratt/pqcrypto-wasi/runs/4330955456?check_suite_focus=true

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,4 +1,7 @@
-on: push
+on:
+  push:
+  pull_request:
+    branches: [master]
```
Comment (thomwiggers) [inline .github/workflows/ci.yml:4]: Should this not run on all PRs.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -12,6 +15,7 @@ jobs:
           - beta
           - nightly
     env:
+      CARGO_TERM_COLOR: always
```
Comment (thomwiggers) [inline .github/workflows/ci.yml:18]: nice

File: WASM.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+Support for WebAssembly
+=======================
+
+# Summary
+
+Due to the need for a standard library in the build these quantum routines it is not
+possible to compile them as pure no-std WASM however by using wasi and its standard
+library it becomes possible.
```
Comment (thomwiggers) [inline WASM.md:8]:
```suggestion
Due to the need for a standard library to build these quantum routines it is not
possible to compile them as pure no-std WASM. However by using wasi and its standard
library, it becomes possible.
```

Why is that, by the way?

File: WASM.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+# Steps
+
+Download the wasm32-wasi sysroot build from https://github.com/WebAssembly/wasi-sdk
+and set the environment variable to point to it
```
Comment (thomwiggers) [inline WASM.md:13]:
```suggestion
Download the wasm32-wasi sysroot build from https://github.com/WebAssembly/wasi-sdk
and set the environment variable to point to it. This example installs it globally and makes it
available system-wide. You may also consider adding it to your profile's environment variables
or exporting the environment variable each time manually.
```

File: WASM.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+cd ..
+wget https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-12/wasi-sysroot-12.0.tar.gz
+tar -xvzf wasi-sysroot-12.0.tar.gz
+rm wasi-sysroot-12.0.tar.gz
+
+sudo tee /etc/profile.d/wasi.sh <<EOF
+export WASI_SDK_DIR="$(pwd)/wasi-sysroot"
+EOF
+source /etc/profile.d/wasi.sh
+````
```
Comment (thomwiggers) [inline WASM.md:25]: This has one backtick too many (I can't make a suggestion here, Github breaks on it)

File: WASM.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+```sh
+cd ..
+wget https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-12/wasi-sysroot-12.0.tar.gz
+tar -xvzf wasi-sysroot-12.0.tar.gz
+rm wasi-sysroot-12.0.tar.gz
+
+sudo tee /etc/profile.d/wasi.sh <<EOF
+export WASI_SDK_DIR="$(pwd)/wasi-sysroot"
+EOF
+source /etc/profile.d/wasi.sh
```
Comment (thomwiggers) [inline WASM.md:24]:
```suggestion
```sh
sudo mkdir -p /opt/wasi
cd /opt/wasi
wget https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-12/wasi-sysroot-12.0.tar.gz -O /tmp/wasi-sysroot-12.0.tar.gz
sudo tar -xvzf /tmp/wasi-sysroot-12.0.tar.gz
rm /tmp/wasi-sysroot-12.0.tar.gz

sudo tee /etc/profile.d/wasi.sh <<EOF
export WASI_SDK_DIR="/opt/wasi/wasi-sysroot"
EOF
source /etc/profile.d/wasi.sh
```


If you're installing something to `profile.d`, better also put it in `/opt` or another global location.

File: WASM.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+````
+
+Note: While the source command brings the WASI export variables into your current
+terminal any new terminals will miss this until you restart your machine.
```
Comment (thomwiggers) [inline WASM.md:28]:
```suggestion
terminal any new terminals will miss this until you log out and back in.
```

**Review by thomwiggers — CHANGES_REQUESTED** [review]
Few nits but this is heading in the right direction.

File: WASM.md (hunk 3 above)
Comment (john-sharratt) [reply WASM.md]: Its quite hard to resolve all the dependencies and while the routines that the code uses don't really use things from the standard library that would prohibit them running in a sandbox environment (such as no-std) unfortunately the dependency graphs do... which means you end up chasing submitting merge requests to everything upstream until its better supported.

The WASI movement lowers the bar on how much needs to be refactored upstream as it keeps some of the common functions everyone is dependent on.

In a way the very problem is the reason I'm submitting this to your code upstream as if I don't its not my dependency that hurts but its the dependency on the dependencies.

Over time this problem will go away as more and more upstream merges are accepted - once we get a critical mass on WebAssembly WASI compatible libraries then I think the community will move onto "no-std" patches more and more.

Basically for me personally I run two binaries for most of my code, one that compiles on
--target wasm32-unknown-unknown (a.k.a. no-std)

And another on
--target wasm32-wasi

That way I can work on different upstream merges based on the health of the ecosystem as a whole. pqcrypto falls in the wasm32-wasi bucket because its largely still written with the standard library which is just too hard to remove entirely... at the moment at least.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (john-sharratt) [reply .github/workflows/ci.yml:18]: looks beautiful hey :-)

File: WASM.md (hunk 5 above)
Comment (john-sharratt) [reply WASM.md]: fixed

File: .github/workflows/ci.yml (hunk 1 above)
Comment (john-sharratt) [reply .github/workflows/ci.yml]: switched this to now work on all pull requests

**Comment by john-sharratt** [comment]
All green.
https://github.com/john-sharratt/pqcrypto-wasi/runs/4379567593?check_suite_focus=true

**Review by thomwiggers — APPROVED** [review]

**Comment by john-sharratt** [comment]
Cool... let me know when you next publish to cargo.io and I'll kill the other packages and update dependencies for downstream projects.
